pandas/plotting/_matplotlib/groupby.py

```diff
@@ -0,0 +1,137 @@
+from typing import (
+    Dict,
+    Optional,
+    Union,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    FrameOrSeriesUnion,
+    IndexLabel,
+)
+
+from pandas.core.dtypes.missing import remove_na_arraylike
+
+from pandas import (
+    DataFrame,
+    MultiIndex,
+    Series,
+    concat,
+)
+
+
+def create_iter_data_given_by(
+    data: DataFrame, kind: str = "hist"
+) -> Dict[str, FrameOrSeriesUnion]:
+    """
+    Create data for iteration given `by` is assigned or not, and it is only
+    used in both hist and boxplot.
+
+    If `by` is assigned, return a dictionary of DataFrames in which the key of
+    dictionary is the values in groups.
+    If `by` is not assigned, return input as is, and this preserves current
+    status of iter_data.
+
+    Parameters
+    ----------
+    data : reformatted grouped data from `_compute_plot_data` method.
+    kind : str, plot kind. This function is only used for `hist` and `box` plots.
+
+    Returns
+    -------
+    iter_data : DataFrame or Dictionary of DataFrames
+
+    Examples
+    --------
+    If `by` is assigned:
+
+    >>> import numpy as np
+    >>> tuples = [('h1', 'a'), ('h1', 'b'), ('h2', 'a'), ('h2', 'b')]
+    >>> mi = MultiIndex.from_tuples(tuples)
+    >>> value = [[1, 3, np.nan, np.nan],
+    ...          [3, 4, np.nan, np.nan], [np.nan, np.nan, 5, 6]]
+    >>> data = DataFrame(value, columns=mi)
+    >>> create_iter_data_given_by(data)
+    {'h1': DataFrame({'a': [1, 3, np.nan], 'b': [3, 4, np.nan]}),
+     'h2': DataFrame({'a': [np.nan, np.nan, 5], 'b': [np.nan, np.nan, 6]})}
+    """
+
+    # For `hist` plot, before transformation, the values in level 0 are
+    # actual subplot titles, and used for column subselection and iteration;
+    # For `box` plot, that's values in level 1
+    if kind == "hist":
+        level = 0
+    elif kind == "box":
+        level = 1
+    else:
+        raise ValueError(
+            f"create_iter_data_given_by can only be used with "
+            f"kind 'hist' and 'box' plots, but used with '{kind}'"
+        )
+
+    iter_data: Dict[str, FrameOrSeriesUnion]
+
+    # Select sub-columns based on the value of first level of MI
+    assert isinstance(data.columns, MultiIndex)
+    cols = data.columns.levels[level]
```

Comment:
Maybe just a personal preference, but to me using the value directly in the comprehension below, instead of assigning it to an intermediate variable, would make things more readable.

Reply:
Sure thing! Like this way too!